src/source-dpdk.c

```diff
@@ -460,27 +460,63 @@ static inline Packet *PacketInitFromMbuf(DPDKThreadVars *ptv, struct rte_mbuf *m
     return p;
 }
 
-static inline void DPDKSegmentedMbufWarning(struct rte_mbuf *mbuf)
+static inline void DPDKSegmentedMbufWarning(void)
 {
     static thread_local bool segmented_mbufs_warned = false;
-    if (!segmented_mbufs_warned && !rte_pktmbuf_is_contiguous(mbuf)) {
-        char warn_s[] = "Segmented mbufs detected! Redmine Ticket #6012 "
-                        "Check your configuration or report the issue";
+    if (unlikely(!segmented_mbufs_warned)) {
         enum rte_proc_type_t eal_t = rte_eal_process_type();
         if (eal_t == RTE_PROC_SECONDARY) {
-            SCLogWarning("%s. To avoid segmented mbufs, "
-                         "try to increase mbuf size in your primary application",
-                    warn_s);
-        } else if (eal_t == RTE_PROC_PRIMARY) {
-            SCLogWarning("%s. To avoid segmented mbufs, "
-                         "try to increase MTU in your suricata.yaml",
-                    warn_s);
+            SCLogWarning("Segmented mbufs detected! Suricata handles them but for better "
+                         "performance, try to increase mbuf size in your primary application");
+        } else {
+            SCLogWarning("Segmented mbufs detected! Suricata handles them but for better "
+                         "performance, try to increase MTU in your suricata.yaml");
         }
-
         segmented_mbufs_warned = true;
     }
 }
 
+/**
+ * \brief Handle segmented (chained) mbufs by linearizing them
+ *
+ * For segmented mbufs, attempts to linearize the data into a contiguous buffer.
+ * First tries rte_pktmbuf_linearize() which copies all segment data into the
+ * first segment. If that fails (not enough tailroom), copies data into the
+ * packet's internal buffer.
+ *
+ * \param p Pointer to the Packet structure
+ * \param mbuf Pointer to the DPDK mbuf
+ * \return 0 on success, -1 on failure
+ */
+static inline int DPDKSegmentedMbufHandle(Packet *p, struct rte_mbuf *mbuf)
+{
+    if (rte_pktmbuf_linearize(mbuf) == 0) {
+        PacketSetData(p, rte_pktmbuf_mtod(mbuf, uint8_t *), rte_pktmbuf_pkt_len(mbuf));
+        return 0;
+    }
+
+    /* Linearization failed (not enough tailroom), copy to packet buffer */
+    uint32_t pkt_len = rte_pktmbuf_pkt_len(mbuf);
+    if (unlikely(pkt_len > MAX_PAYLOAD_SIZE)) {
+        SCLogWarning("Segmented mbuf too large to copy: %u > %d", pkt_len, MAX_PAYLOAD_SIZE);
```

Comment:
"Warning once" is the preferred approach in capture methods to not flood users with warnings/errors in incorrectly configured deployments.